cpp/src/sampling/detail/graph_functions.hpp

```diff
@@ -0,0 +1,176 @@
+ *
+ * Calculate local degree and perform row wise exclusive scan over all gpus in column
+ * communicator.
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return Tuple of two device vectors. The first one contains per source edge-count encountered
+ * by gpus in the column communicator before current gpu. The second device vector contains the
+ * global out degree for every source represented by current gpu
+ */
+template <typename GraphViewType>
+std::tuple<rmm::device_uvector<typename GraphViewType::edge_type>,
+           rmm::device_uvector<typename GraphViewType::edge_type>>
+get_global_degree_information(raft::handle_t const& handle, GraphViewType const& graph_view);
+
+/**
+ * @brief Calculate global adjacency offset for all majors represented by current gpu
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param[in] global_degree_offsets Global degree offset to local adjacency list for every major
+ * represented by current gpu
+ * @param global_out_degrees Global out degrees for every source represented by current gpu
+ * @return Device vector containing the number of edges that are prior to the adjacency list of
+ * every major that can be represented by the current gpu
+ */
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> get_global_adjacency_offset(
+  raft::handle_t const& handle,
+  GraphViewType const& graph_view,
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_degree_offsets,
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_out_degrees);
+
+/**
+ * @brief Gather active majors across gpus in a column communicator
+ *
+ * Collect all the vertex ids and client gpu ids to be processed by every gpu in
+ * the column communicator and sort the list.
+ *
+ * @tparam vertex_t Type of vertex indices.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param d_in Device vector containing vertices local to this GPU
+ * @return Device vector containing all the vertices that are to be processed by every gpu
+ * in the column communicator
+ */
+template <typename vertex_t>
+rmm::device_uvector<vertex_t> gather_active_majors(raft::handle_t const& handle,
+                                                   rmm::device_uvector<vertex_t>&& d_in);
+
+/**
+ * @brief Return global out degrees of active majors
```

Comment:
Need to double check "out" degrees here is correct.

Reply:
Sampling forces `store_transposed=false`